### DISCUSSION TITLE
Some option to bypass the remapping of localhost

### DISCUSSION THREAD

**Discussion by stchristian** [invertase/react-native-firebase#5816] [Ideas]
I came accross a problem with emulators and the way they are handling `localhost` addresses, described [here](https://github.com/invertase/react-native-firebase/issues/4810).

It's nice that now remapping is logged out, but it would be nicer to bypass this feature with some flag, so I don't have to always reconfigure Host IP when I am connecting to another network.
I would love to see [this](https://reactnative.dev/docs/running-on-device) work with firebase emulators, when running and testing apps on real devices.

**Comment by mikehardy** [comment]
Hi there! I'd be open to a PR that implemented that somehow. Perhaps a firebase.json key? 

It's possible in javascript to get the full contents of the firebase.json, so fetching the JSON then that key (if available) may be used as a conditional in the remapping code?

Something like this I think ?

https://github.com/invertase/react-native-firebase/blob/master/packages/app/e2e/config.e2e.js#L28-L35

**Comment by stchristian** [comment]
Yes, a key in the firebase.json would be great. Is there any example when this config is used in JS? 
I'd be happy to open a PR in the future

**Reply by stchristian** [reply]
@mikehardy 

I noticed all modules extend the `FirebaseModule` and it has a [getter](https://github.com/invertase/react-native-firebase/blob/c94546d8127606dca5bfd09ef92ec32eec333f19/packages/app/lib/internal/FirebaseModule.js#L35-L41) for firebase.json.

I guess than I could do something like
```js
    const bypassUrlRemap = this.firebaseJson.bypass_url_remap
    if (!bypassUrlRemap && isAndroid && _url) {
      if (_url.startsWith('http://localhost')) {
        _url = _url.replace('http://localhost', 'http://10.0.2.2');
        // eslint-disable-next-line no-console
        console.log(
          'Mapping auth host "localhost" to "10.0.2.2" for android emulators. Use real IP on real devices.',
        );
      }
```

**Reply by mikehardy** [reply]
ah yes! Much cleaner, that's what I was thinking was possible but I couldn't find a code reference for it, you're handier than me with the API there ;-)

The vast majority of everyone will not have that key defined so please take extra care to make sure the current default (localhost is remapped) is still the way it works if people put invalid / non-boolean-etc values in there or it's missing

**Reply by stchristian** [reply]
I am on this issue, but need a little help.
I did the following in function, auth, firestore modules:
```js
    const androidBypassEmulatorUrlRemap =
      typeof this.firebaseJson.android_bypass_emulator_url_remap === 'boolean' &&
      this.firebaseJson.android_bypass_emulator_url_remap;
    if (!androidBypassEmulatorUrlRemap && isAndroid && _origin) {
```

I tried to run tests (`yarn tests:jest`) but it throws error when the above mentioned _getter_ runs, it says it couldn't parse the json file. I logged out the arguments that it got and its `undefined`... I guess I need to somehow initalize a native module but don't really know how?!

UPDATE:
Okay, found out that I could provide a firebase.json mock in `jest.setup.ts` for the JS only tests. However it would be nice to somehow test this behaviour...

**Reply by mikehardy** [reply]
I see your update - yes you'll have to mock it, for an on device test add it to the e2e javascript tests you see in the packages and then run that harness - it's not as dire as it sounds, especially since it's android-only for the test and you can use a `.only` while writing the test in combo with `yarn tests:android:test-reuse` to really quickly iterate - https://github.com/invertase/react-native-firebase/blob/master/tests/README.md has all the instructions

**Reply by mikehardy** [reply]
a new test after this one for instance might serve ? https://github.com/invertase/react-native-firebase/blob/514e6bd51e7624a6403dda706f4e5b65cee63422/packages/app/e2e/config.e2e.js#L28-L34

**Reply by stchristian** [reply]
Finally I've got to run the tests. It was a nice experience, however I don't think I should add tests for this flag.
I guess I'll just simply extend the documentation

**Reply by mikehardy** [reply]
ok

**Comment by stchristian** [comment]
Hi again!

I got the code working with the patches, but it took a lot of time to figure out that the `android_bypass_emulator_url_remap` should be under a `react-native` key in the firebase.json in order for the flag to work.

Is this the expected behaviour?
If so the [documentation](https://rnfirebase.io/app/json-config) should really say something about that

**Reply by mikehardy** [reply]
A reasonable point, just added this https://github.com/invertase/react-native-firebase/commit/11624756989ccd119dac6824e464721bd954594e
It'll publish on the page you linked as soon as Vercel finishes chewing on it and deploys, some 5-10mins
